core/src/poh_service.rs

```diff
@@ -254,24 +259,38 @@ impl PohService {
                     timing.num_hashes += hashes_per_batch;
                     let mut hash_time = Measure::start("hash");
                     let should_tick = poh_l.hash(hashes_per_batch);
+                    let ideal_time = poh_l.target_poh_time(target_ns_per_tick);
                     hash_time.stop();
                     timing.total_hash_time_ns += hash_time.as_ns();
                     if should_tick {
                         // nothing else can be done. tick required.
                         return true;
                     }
                     // check to see if a record request has been sent
-                    let get_again = record_receiver.try_recv();
-                    match get_again {
-                        Ok(record) => {
+                    if let Ok(record) = record_receiver.try_recv() {
+                        // remember the record we just received as the next record to occur
+                        *next_record = Some(record);
+                        break;
+                    }
+                    // check to see if we need to wait to catch up to ideal
+                    let wait_start = Instant::now();
+                    if ideal_time <= wait_start {
+                        // no, keep hashing. We still hold the lock.
+                        continue;
+                    }
+
+                    // busy wait, polling for new records and after dropping poh lock (reset can occur, for example)
+                    drop(poh_l);
+                    while ideal_time > Instant::now() {
+                        // check to see if a record request has been sent
+                        if let Ok(record) = record_receiver.try_recv() {
                             // remember the record we just received as the next record to occur
                             *next_record = Some(record);
                             break;
                         }
-                        Err(_) => {
-                            continue;
```

Comment:
Oh good catch, would this have looped forever if the `record_receiver` was dropped before (such that this always errored?)

Reply:
I don't think this change significantly changes what happens on an error - especially if all record_senders are dropped. I think we'll always get a try_recv() error and we'll always just ignore the error and keep hashing until it is time to tick and then we notice pohservice is told to quit.

Reply:
ah yeah, there's a return in the `if should_tick {`